includes/polling/functions.inc.php

```diff
@@ -690,3 +690,75 @@ function convert_to_celsius($value)
     $value = ($value - 32) / 1.8;
     return sprintf('%.02f', $value);
 }
+
+
+/**
+ * This is to make it easier polling apps. Also to help standardize around JSON.
+ * @param $device
+ * @param string $poin partial OID
+ *
+ * In regards to the POID, if the OID is 'nsExtendOutputFull.3.122.102.115', then
+ * it would become '3.122.102.115'.
+ *
+ * The returned value is JSON parsed into a array.
+ *
+ * The key error contains the any errorcode and 0 if non is present. A textual
+ * description of said error is found in errorString . The version of the return
+ * can be found in version.
+ *
+ * Possible parsing related errors.
+ * -2 : Could not decode the JSON.
+ * -3 : Empty JSON parsed, meaning blank JSON was returned.
+ * -4 : Empty return from snmp_get().
+ * -5 : Non-numeric version number.
+ *
+ * Other error codes are possible, but those are extend related and should be handled
+ * by the poller calling this function.
+ *
+ * If version, error, and errorString are not set, it is assumed it is a old JSON backend and
+ * sets them manually with a version of 0 and assumes a error of 0(no error).
+ */
+function json_app_get($device, $poid)
+{
+    $returned_json = snmp_get($device, 'nsExtendOutputFull.'.$poid, '-O qv', 'NET-SNMP-EXTEND-MIB');
+
+    // make sure we actually get something back
+    if (empty($returned_json)) {
+        $parsed_json=json_decode('{"error":"-4","errorString":"Empty return from snmp_get()"}', true);
+        return $parsed_json;
+    }
+
+    $parsed_json=json_decode(stripslashes($returned_json), true);
+
+    if (json_last_error() === JSON_ERROR_NONE) {
+        if (empty($parsed_json)) {
+            // If we get here it means there are no keys in the array, meaining '{}' was was returned
+            $parsed_json{'error'}='-3';
+            $parsed_json{'errorString'}='Blank JSON returned.';
+        } else {
+            // If the following is true, it is a legacy JSON app extend, meaning these are not set
```

Comment:
You should probably return the raw output so we can process legacy apps without another snmpget.

Reply:
This does not cover non-JSON stuff. That will require converting the backends and updating the pollers.

The legacy stuff AFAIK at this point is only ZFS as those variables are not present in the returned JSON.

While also for that, it prevents possibly breakages from happening if the backend errors and does not generates valid JSON, but not with some of the expected and commonly used keys.

Actually while I am thinking of this, I should add in a third option for this all for minimum return version, causing it to error when that happens.